### PR TITLE
Schema components further fixes and improvements

### PR DESCRIPTION
**Pull request by E-gy**
TLDR: Follow-up to #61, improving behaviour for generic impls over component subtypes

1. Schema for `Option<T>` where `T` is a component now uses `nullable` + `oneOf` linking to `T`'s component.
One _has_ to go through `oneOf` direction because if `$ref` is present at root, all other properties are ignored.
2. For `[T]`&friends fixes a bug where generated component would nest `T`'s schema directly in `items` instead of `$ref`ing to it
3. Following up #66
    1. provided additional `Map` friends implementations (`BTreeMap` & `IndexMap`), and additionally over `Cow<String>`.
    2. if `T` is a component, (just as for `Option` and `[]`,) generate component for the map with `_Map` suffix and refer to the component in `additionalProperties` 
4. Persistent component name for primitives accounts for nullability (thus usage of `Option<u64>` resulting in `u64_Opt` component where requested).
4. Extended tests for `HashMap` (components) as well as to testing use of `$ref`

### CODE REVIEW

**Comment by E-gy** [comment]
For 1st change, refer to https://github.com/OAI/OpenAPI-Specification/issues/1368.
TLDR: OpenAPI 3.0 there isn't _the good way_. In OpenAPI 3.1 instead of `nullable` (which is removed), `oneOf: [{ type: "null" },  { $ref: ... }]` should be used

**Review by kdy1 — APPROVED** [review]
LGTM. Thanks!
